mpf/devices/blinkenlight.py

```diff
@@ -0,0 +1,164 @@
+        super().load_config(config)
+
+        self._color_duration = self.config['color_duration']
+        self._cycle_duration = self.config['cycle_duration']
+        if (self._color_duration is None and self._cycle_duration is None) or \
+           (self._color_duration is not None and self._cycle_duration is not None):
+            self._blinkenlight_validation_error(
+                "Either color_duration or cycle_duration must be specified, but not both.", 1)
+
+    def _blinkenlight_validation_error(self, msg, error_code) -> "NoReturn":  # pragma: no cover
+        raise ConfigFileError('"{}" >> {}'.format(self.name, msg), error_code, "blinkenlight", self.name)
+
+    @property
+    def num_colors(self):
+        """Return the number of colors (excluding "off") for this blinkenlight."""
+        return self._num_colors
+
+    @num_colors.setter
+    def num_colors(self, newvalue):
+        self._num_colors = newvalue
+        # if the number of colors is 0, remove us from the light stack
+        if self._num_colors == 0:
+            self.light.remove_from_stack_by_key(self._blinkenlight_key)
+
+    @property
+    def light(self):
+        """Return the light this blinkenlight controls."""
+        return self.config['light']
+
+    @property
+    def num_colors_in_cycle(self):
+        """Similar to num_colors, but adds 1 for the "off" color if there is one between cycles."""
+        return self.num_colors + (1 if self._off_between_cycles() else 0)
+
+    @property
+    def _blinkenlight_key(self):
+        # this is the key that's passed to the light as the color's key
+        # so that we can remove this blinkenlight's color from the light stack
+        # when all the blinkenlight's colors are removed.
+        # This prevents the light from staying in the off state if the blinkenlight
+        # was on top of the priority list.
+        return 'blinkenlight_{}'.format(self.name)
+
+    def add_color(self, color, label, priority, context):
+        """Add a color to the blinkenlight."""
+        # check if this label already exists. If it does, replace it with the incoming color/priority
+        existing_color = [x for x in self._colors if x[1] == label]
+        if len(existing_color) == 0:
+            self._colors.append((color, label, priority, context))
+            self.num_colors += 1
+            self.info_log('Color {} with label {} added'.format(color, label))
+            self._restart()
+        elif len(existing_color) == 1:
+            # color with this label already exists. Just update it with this new color and priority
+            self._remove_color_with_label(label)
+            self.add_color(color, label, priority, context)
+
+    def remove_all_colors(self):
+        """Remove all colors from the blinkenlight."""
+        self._colors.clear()
+        self.num_colors = 0
+        self.info_log('All colors removed')
+        self._restart()
+
+    def remove_color_with_label(self, label):
```

Comment:
This needs changed to private _remove_color_with_label (or if it needs to be public, line 92 needs changed to remove the underscore.

Reply:
Blinkenlight_player calls this, so it looks like the call on 92 needs to be global.

Reply:
Good catch.  That would have been a runtime error.  Made me realize I didn't have a unit test for this.  I've fixed it and added a test.  Will push in next update.